File: scripts/scrape_masterplans.js
#!/usr/bin/env node
/*
  Planora scraper (developer websites first)
  - Reads config/sources.json
  - Fetches page html
  - Extracts title and og:image
  - Downloads image to assets/scraped
  - Upserts entry in data/properties.json

  Notes:
  - Respect each website's Terms of Use and robots policy.
  - Some pages are JS-rendered; those may need a headless browser later.
*/

const fs = require("fs");
const path = require("path");
const crypto = require("crypto");

const ROOT = path.join(__dirname, "..");
const DB_PATH = path.join(ROOT, "data", "properties.json");
const SRC_PATH = path.join(ROOT, "config", "sources.json");
const ASSET_DIR = path.join(ROOT, "assets", "scraped");

if (!fs.existsSync(ASSET_DIR)) fs.mkdirSync(ASSET_DIR, { recursive: true });

function slugify(text) {
  return String(text || "")
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "")
    .slice(0, 80);
}

function loadJson(filePath) {
  return JSON.parse(fs.readFileSync(filePath, "utf-8"));
}

function saveJson(filePath, data) {
  fs.writeFileSync(filePath, JSON.stringify(data, null, 2) + "\n", "utf-8");
}

function parseTitle(html) {
  const m = html.match(/<title[^>]*>([\s\S]*?)<\/title>/i);
  return m ? m[1].trim().replace(/\s+/g, " ") : "Unknown Plan";
}

function parseMeta(html, name) {
  const rx = new RegExp(
    `<meta[^>]+(?:property|name)=["']${name}["'][^>]+content=["']([^"']+)["'][^>]*>`,
    "i"
  );
  const m = html.match(rx);
  return m ? m[1].trim() : "";
}

function absolutizeUrl(baseUrl, maybeRelative) {
  try {
    return new URL(maybeRelative, baseUrl).toString();
  } catch {
    return "";
  }
}

async function fetchText(url) {
  const res = await fetch(url, {
    headers: {
      "user-agent": "PlanoraBot/1.0 (+https://planora.bck.life)",
      accept: "text/html,application/xhtml+xml"
    }
  });
  if (!res.ok) throw new Error(`Failed ${url} (${res.status})`);
  return await res.text();
}

async function downloadFile(url, localPath) {
  const res = await fetch(url, {
    headers: { "user-agent": "PlanoraBot/1.0 (+https://planora.bck.life)" }
  });
  if (!res.ok) throw new Error(`Image failed ${url} (${res.status})`);
  const arr = await res.arrayBuffer();
  fs.writeFileSync(localPath, Buffer.from(arr));
}

function guessExt(url) {
  const clean = url.split("?")[0].toLowerCase();
  if (clean.endsWith(".png")) return ".png";
  if (clean.endsWith(".webp")) return ".webp";
  if (clean.endsWith(".jpeg")) return ".jpeg";
  return ".jpg";
}

function inferNameFromTitle(title) {
  const first = title.split("|")[0].split("-")[0].trim();
  return first.length > 2 ? first : "Master Plan";
}

function upsertProperty(db, entry) {
  const idx = db.properties.findIndex((p) => p.id === entry.id);
  if (idx >= 0) db.properties[idx] = { ...db.properties[idx], ...entry };
  else db.properties.push(entry);
}

async function run() {
  const db = loadJson(DB_PATH);
  const sources = loadJson(SRC_PATH).sources || [];

  for (const src of sources) {
    try {
      const html = await fetchText(src.url);
      const title = parseTitle(html);
      const ogImage = parseMeta(html, "og:image");
      const fullImageUrl = ogImage ? absolutizeUrl(src.url, ogImage) : "";

      const inferredName = inferNameFromTitle(title);
      const idBase = slugify(inferredName || src.community || src.developer || "plan");
      const hash = crypto.createHash("md5").update(src.url).digest("hex").slice(0, 6);
      const id = `${idBase}-${hash}`;

      let imagePath = "";
      if (fullImageUrl) {
        const ext = guessExt(fullImageUrl);
        const localFile = `${id}${ext}`;
        const absLocal = path.join(ASSET_DIR, localFile);
        await downloadFile(fullImageUrl, absLocal);
        imagePath = `assets/scraped/${localFile}`;
      }

      const entry = {
        id,
        name: src.name || inferredName,
        developer: src.developer || "Unknown Developer",
        community: src.community || "Unknown Community",
        city: src.city || "Dubai",
        country: src.country || "UAE",
        type: src.type || "Townhouse",
        beds: src.beds || "N/A",
        image: imagePath || "assets/generated/hero-illustration.svg",
        affiliateUrl: src.url,
        sourceUrl: src.url,
        sourceTitle: title,
        rooms: src.rooms || {}
      };

      upsertProperty(db, entry);
      console.log(`Upserted: ${entry.name} (${entry.id})`);
    } catch (err) {
      console.error(`Source failed: ${src.url}`);
      console.error(String(err.message || err));
    }
  }

  saveJson(DB_PATH, db);
  console.log("Done. Database updated:", DB_PATH);
}

run();

File: scripts/scrape_masterplans.py
#!/usr/bin/env python3
"""
Planora scraper in Python.
- Reads config/sources.json
- Fetches developer pages
- Extracts <title> and og:image
- Downloads image to assets/scraped
- Upserts entries in data/properties.json
"""

from __future__ import annotations

import hashlib
import json
import re
from pathlib import Path
from urllib.parse import urljoin, urlparse
from urllib.request import Request, urlopen
from urllib.error import URLError, HTTPError

ROOT = Path(__file__).resolve().parent.parent
DB_PATH = ROOT / "data" / "properties.json"
SRC_PATH = ROOT / "config" / "sources.json"
ASSET_DIR = ROOT / "assets" / "scraped"

ASSET_DIR.mkdir(parents=True, exist_ok=True)

USER_AGENT = "PlanoraBot/1.0 (+https://planora.bck.life)"


def slugify(text: str) -> str:
    text = (text or "").lower()
    text = re.sub(r"[^a-z0-9]+", "-", text)
    return text.strip("-")[:80] or "plan"


def load_json(path: Path) -> dict:
    return json.loads(path.read_text(encoding="utf-8"))


def save_json(path: Path, data: dict) -> None:
    path.write_text(json.dumps(data, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")


def fetch_bytes(url: str) -> bytes:
    req = Request(url, headers={"User-Agent": USER_AGENT, "Accept": "text/html,application/xhtml+xml,*/*"})
    with urlopen(req, timeout=30) as resp:
        return resp.read()


def fetch_text(url: str) -> str:
    data = fetch_bytes(url)
    return data.decode("utf-8", errors="replace")


def parse_title(html: str) -> str:
    m = re.search(r"<title[^>]*>([\s\S]*?)</title>", html, flags=re.IGNORECASE)
    return re.sub(r"\s+", " ", m.group(1)).strip() if m else "Unknown Plan"


def parse_og_image(html: str) -> str:
    patterns = [
        r'<meta[^>]+property=["\']og:image["\'][^>]+content=["\']([^"\']+)["\'][^>]*>',
        r'<meta[^>]+content=["\']([^"\']+)["\'][^>]+property=["\']og:image["\'][^>]*>',
        r'<meta[^>]+name=["\']og:image["\'][^>]+content=["\']([^"\']+)["\'][^>]*>',
    ]
    for p in patterns:
        m = re.search(p, html, flags=re.IGNORECASE)
        if m:
            return m.group(1).strip()
    return ""


def infer_name_from_title(title: str) -> str:
    part = title.split("|")[0].split("-")[0].strip()
    return part if len(part) > 2 else "Master Plan"


def guess_ext(image_url: str) -> str:
    path = urlparse(image_url).path.lower()
    if path.endswith(".png"):
        return ".png"
    if path.endswith(".webp"):
        return ".webp"
    if path.endswith(".jpeg"):
        return ".jpeg"
    return ".jpg"


def upsert_property(db: dict, entry: dict) -> None:
    props = db.get("properties", [])
    for i, prop in enumerate(props):
        if prop.get("id") == entry["id"]:
            props[i] = {**prop, **entry}
            return
    props.append(entry)


def run() -> None:
    db = load_json(DB_PATH)
    sources = load_json(SRC_PATH).get("sources", [])

    for src in sources:
        url = src.get("url", "").strip()
        if not url:
            continue
        try:
            html = fetch_text(url)
            title = parse_title(html)
            og_image = parse_og_image(html)
            full_image_url = urljoin(url, og_image) if og_image else ""

            inferred_name = infer_name_from_title(title)
            id_base = slugify(inferred_name or src.get("community") or src.get("developer") or "plan")
            hash6 = hashlib.md5(url.encode("utf-8")).hexdigest()[:6]
            prop_id = f"{id_base}-{hash6}"

            image_rel = "assets/generated/hero-illustration.svg"
            if full_image_url:
                ext = guess_ext(full_image_url)
                local_name = f"{prop_id}{ext}"
                local_path = ASSET_DIR / local_name
                local_path.write_bytes(fetch_bytes(full_image_url))
                image_rel = f"assets/scraped/{local_name}"

            entry = {
                "id": prop_id,
                "name": src.get("name") or inferred_name,
                "developer": src.get("developer", "Unknown Developer"),
                "community": src.get("community", "Unknown Community"),
                "city": src.get("city", "Dubai"),
                "country": src.get("country", "UAE"),
                "type": src.get("type", "Townhouse"),
                "beds": src.get("beds", "N/A"),
                "image": image_rel,
                "affiliateUrl": url,
                "sourceUrl": url,
                "sourceTitle": title,
                "rooms": src.get("rooms", {}),
            }

            upsert_property(db, entry)
            print(f"Upserted: {entry['name']} ({entry['id']})")
        except HTTPError as e:
            print(f"Source failed: {url} -> HTTP {e.code}")
        except URLError as e:
            print(f"Source failed: {url} -> URL error: {e.reason}")
        except Exception as e:
            print(f"Source failed: {url} -> {e}")

    save_json(DB_PATH, db)
    print(f"Done. Database updated: {DB_PATH}")


if __name__ == "__main__":
    run()
